### DISCUSSION TITLE
lifespan incompatible type

### DISCUSSION THREAD

**Discussion by dineshbvadhia** [fastapi/fastapi#10413] [Questions] [answered]
### First Check

- [X] I added a very descriptive title here.
- [X] I used the GitHub search to find a similar question and didn't find it.
- [X] I searched the FastAPI documentation, with the integrated search.
- [X] I already searched in Google "How to X in FastAPI" and didn't find any information.
- [X] I already read and followed all the tutorial in the docs and didn't find an answer.
- [X] I already checked if it is not related to FastAPI but to [Pydantic](https://github.com/pydantic/pydantic).
- [X] I already checked if it is not related to FastAPI but to [Swagger UI](https://github.com/swagger-api/swagger-ui).
- [X] I already checked if it is not related to FastAPI but to [ReDoc](https://github.com/Redocly/redoc).

### Commit to Help

- [X] I commit to help with one of those options 👆

### Example Code

```python
async def lifespan(_: FastAPI) ->  AsyncGenerator[dict[str, str] | None, None]:
    async with CustomAsyncContextManager() as c:
	# startup stuff
	yield {'this' : that}
	# cleanup

app = FastAPI(lifespan=lifespan)	# L866
```

### Description

I'm using a custom async context manager with lifespan. The program runs as expected but mypy returns this error:

mypy:
`866: error: Argument "lifespan" to "FastAPI" has incompatible type "Callable[[FastAPI], AsyncGenerator[dict[str, Dao] | None, None]]"; expected "Callable[[FastAPI], AbstractAsyncContextManager[None]] | Callable[[FastAPI], AbstractAsyncContextManager[Mapping[str, Any]]] | None"  [arg-type]`

What does this mean?

### Operating System

Windows

### FastAPI Version

0.103.2

### Pydantic Version

2.4.2

### Python Version

Python 3.11.4

**Comment by dineshbvadhia** [comment]
@Kludex Hi. What is the answer to this problem? The background is mypy passed the codebase successfully but mypyc faulted in the lifespan code with a "async generators are unimplemented" error. It was suggested to replace the @asynccontextmanager with a separate simple Async Context Manager. Hope this helps.

**Reply by dineshbvadhia** [reply]
Btw, the code for the CustomAsyncContextManager is:
```
class CustomAsyncContextManager:
    async def __aenter__(self) -> None:
        return

    async def __aexit__(self, exc_type: type[BaseException], exc_value: BaseException, exc_traceback: TracebackType | None) -> None:
        return
```

**Comment by jgould22** [comment]
What happens if you rewrite your lifespan implementation to resemble the [starlette example](https://www.starlette.io/lifespan/) ?

**Reply by dineshbvadhia** [reply]
@jgould22 The Lifespan or the Lifespan-State example?

**Reply by jgould22** [reply]
Since you appear to be trying to return that dict I would assume the state example. FastAPI only has examples with an empty yield in the docs.

**Reply by dineshbvadhia** [reply]
Something like this:
```
class Dao:
    async def __aenter__(self):
        dao = blah()
	return dao
    async def __aexit__(self):
	return

class State(TypedDict):
    dao: Dao

@contextlib.asynccontextmanager
async def lifespan(app: Starlette) -> AsyncIterator[State]:
    async with Dao() as dao:
        # startup stuff
        yield {'dao': dao}
	# cleanup stuff

app = FastAPI(lifespan=lifespan)
```
If so, won't the `yield` cause a "async generators are unimplemented" error with `mypyc`?

**Reply by dineshbvadhia** [reply]
I implemented this in my codebase and `mypyc` faulted on the `yield` line with `"async generators are unimplemented"`.

**Reply by YuriiMotov** [reply, marked as answer]
Correct type annotations were needed:

```python
from collections.abc import AsyncIterator
import contextlib
from typing import TypedDict, cast

from fastapi import FastAPI, Request


class Blah:
    pass


class Dao:
    async def __aenter__(self):
        dao = Blah()
        return dao

    async def __aexit__(self, exc_type, exc, tb):
        return


class State(TypedDict):
    dao: Dao


@contextlib.asynccontextmanager
async def lifespan(app: FastAPI) -> AsyncIterator[State]:
    async with Dao() as dao:
        # startup stuff
        yield {"dao": dao}
        # cleanup stuff


app = FastAPI(lifespan=lifespan)


@app.get("/")
async def get_(request: Request):
    assert isinstance(cast(State, request.state._state)["dao"], Blah)

```
